### DISCUSSION TITLE
How to automatically signin with query params

### DISCUSSION THREAD

**Discussion by muuhoffman** [nextauthjs/next-auth#2907] [Help]
### Question 💬

I am using a custom credentials to sign-in with a token passed into my web app via query params. 

```javascript
// index.tsx

export const getServerSideProps: GetServerSideProps = async function (context) {
  const session = await getSession(context);
  if (!session && context.query.public_token) {
    // automatically log in user here
  } else if (!session) {
    return {
      redirect: {
        destination: "/api/auth/signin",
        permanent: false,
      },
    };
  }

  return {
    props: { session },
  };
};
```

Is it possible to signin to my custom credentials provider there? I tried calling `/api/auth/signin/credentials` based on the [REST API Documentation](https://next-auth.js.org/v3/getting-started/rest-api#post-apiauthsigninprovider)
```
// NEXTAUTH_URL=https://localhost:3000/api/auth/
const {data} = await axios.post(
      `https://localhost:3000/api/auth/signin/credentials`,
      {
        public_token: context.query.public_token,
      },
      {
        headers: {
          Accept: "application/json",
          "Content-Type": "application/json",
        },
      }
    );
```

but `data` is just returning the html for my signin page.

### How to reproduce ☕️

```
export const getServerSideProps: GetServerSideProps = async function (context) {
  const session = await getSession(context);
  if (!session && context?.query?.public_token) {
    const { data } = await axios.post(
      `${process.env.NEXTAUTH_URL}signin/credentials`,
      {
        public_token: context.query.public_token,
      },
      {
        headers: {
          Accept: "application/json",
          "Content-Type": "application/json",
        },
      }
    );
    console.log("Index Sign In", data);
  } else if (!session) {
    return {
      redirect: {
        destination: "/api/auth/signin",
        permanent: false,
      },
    };
  }

  return {
    props: { session },
  };
};
```

### Contributing 🙌🏽

Yes, I am willing to help answer this question in a PR

**Comment by balazsorban44** [comment]
That endpoint is only meant for OAuth providers, as you can see in the docs where you linked:

> Starts an OAuth signin flow for the specified provider.

For the credentials, try sending that request to

`/api/auth/callback/credentials`

Although you won't get the actual session back, but a Set-Cookie header.

See:

https://github.com/nextauthjs/next-auth/blob/e8a58a01b6b8c456268771f9b382fa300ccab8ac/src/client/index.js#L193-L216

https://github.com/nextauthjs/next-auth/blob/e8a58a01b6b8c456268771f9b382fa300ccab8ac/src/server/routes/callback.js#L407

**Reply by alessandrojcm** [reply]
Hi there, did this change on v4? I'm trying to sign in for a credentials provider through the API but all I get as a response is the sign in `url`. This is the request (via `curl`), the `/api/auth/csrf` is called first to get the CSRF Token: 

```shell
curl --request POST \
  --url http://localhost:3000/api/auth/callback/impersonation \
  --header 'Content-Type: application/x-www-form-urlencoded' \
  --cookie 'next-auth.csrf-token=7f8bbe617f9cf4ba618e4c53a5df8a65bf73a58c50ab91689a2dadd61517f90c%257Cf4bb38e3b957006469573fe9faa3c331c30735cfe77341c75fbc5a4ff545190e; next-auth.callback-url=http%253A%252F%252Flocalhost%253A3000%252Fonboarding' \
  --data json=true \
  --data csrfToken=e4282de83ee8307d05f5adcaa9af1dab3d0024505d7ae71ac207e03eb542571a \
  --data email=example@email.com \
  --data token=ExamplePassword \
  --data callbackUrl=http://localhost:3000/onboarding
  ```
  
  What I get as a response is:
  ```json
  {
	"url": "http://localhost:3000/api/auth/signin?csrf=true"
}
```

**Reply by alessandrojcm** [reply]
Nevermind my bad, I was using Insomnia's request chaining to make a `POST` request to the CSRF endpoint at the same time the sign in was made; when I request the token first and add it to the callback request then it works.

**Reply by thierryclipboardhealth** [reply]
Hi @alessandrojcm I'm having the same problem right now. Do you have more information on how you fixed it? I followed exactly the flow that @balazsorban44 suggested and I'm also receiving `{ url: 'http://localhost:3000/api/auth/signin?csrf=true' }` 

Thank you!

**Reply by alessandrojcm** [reply]
Hey @thierryclipboardhealth sorry this was long ago and I don't work at that company anymore so I don't have access to the code. But iirc you need to first request a CSRF token from `next-auth` with `/api/auth/csrf` and then call the sign-in endpoint while including that token in the `next-auth.csrf-token` header.

**Reply by thierryclipboardhealth** [reply]
Ok yeah took a chance! 😅 Thank you for the reply! Will try this. It helps 🙏
